app/js/lib/foxbox/services.js

```diff
@@ -138,91 +136,66 @@ export default class Services extends EventDispatcher {
-          );
+  [p.updateServiceList]() {
+    return Promise.all([
+      this[p.api].get('services'),
+      this[p.db].getServices(),
+      this[p.getCache](),
+    ])
+    .then(([fetchedServices, storedServices, cache]) => {
+      let servicesToAddCount = 0;
+      fetchedServices.forEach((fetchedService) => {
+        const storedService = storedServices.find(
+          (storedService) => storedService.id === fetchedService.id
+        );
+
+        const isExistingService = !!storedService;
+
+        if (isExistingService && isSimilar(fetchedService, storedService)) {
+          return;
+        }
 
-          const isExistingService = !!storedService;
+        // Populate the db with the latest service.
+        this[p.db].setService(fetchedService);
 
-          if (isExistingService && isSimilar(fetchedService, storedService)) {
-            return;
-          }
+        const service = this[p.getServiceInstance](fetchedService);
+        cache.set(service.id, service);
+
+        if (isExistingService) {
+          this.emit('service-changed', service);
+        } else {
+          servicesToAddCount++;
+        }
+      });
 
-          // Populate the db with the latest service.
-          this[p.db].setService(fetchedService);
+      const servicesToRemoveCount = storedServices.length +
+        servicesToAddCount - fetchedServices.length;
+      if (servicesToRemoveCount > 0) {
+        storedServices.forEach((storedService) => {
+          const fetchedService = fetchedServices.find(
+            (fetchedService) => fetchedService.id === storedService.id
+          );
 
-          const service = this[p.getServiceInstance](fetchedService);
-          cache.set(service.id, service);
+          if (!fetchedService) {
+            this[p.db].deleteService(storedService);
 
-          if (isExistingService) {
-            this.emit('service-changed', service);
-          } else {
-            servicesToAddCount++;
+            // We should teardown service instance and remove it from the cache.
+            const cachedService = cache.get(storedService.id);
+            cachedService.teardown();
+            cache.delete(cachedService.id);
           }
         });
+      }
 
-        const servicesToRemoveCount = storedServices.length +
-          servicesToAddCount - fetchedServices.length;
-        if (servicesToRemoveCount > 0) {
-          storedServices.forEach((storedService) => {
-            const fetchedService = fetchedServices.find(
-              (fetchedService) => fetchedService.id === storedService.id
-            );
-
-            if (!fetchedService) {
-              this[p.db].deleteService(storedService);
-              cache.delete(storedService.id);
-            }
-          });
-        }
-
-        if (servicesToAddCount > 0 || servicesToRemoveCount > 0) {
-          this.emit('services-changed');
-        }
-      });
+      if (servicesToAddCount > 0 || servicesToRemoveCount > 0) {
```

Comment:
As I said before, eventually I'd like us to generate two events instead of this one - ```service-added``` with new service as argument and ```service-removed``` with removed service as argument - so that view can unsubscribe from its event handlers safely even still being active and let user know that currently opened service has disappeared.